src/signer/noscrypt_signer.cpp

```diff
@@ -0,0 +1,569 @@
+
+    NC_LOG_ERROR(pubkeyGenerationResult);
+};
+
+#pragma endregion
+
+#pragma region Constructors and Destructors
+
+NoscryptSigner::NoscryptSigner(
+    shared_ptr<plog::IAppender> appender,
+    shared_ptr<INostrServiceBase> nostrService
+)
+{
+    plog::init(plog::debug, appender.get());
+
+    this->_noscryptContext = initNoscryptContext();
+
+    createLocalKeypair(
+        this->_noscryptContext, 
+        this->_localPrivateKey, 
+        this->_localPublicKey
+    );
+
+    this->_nostrService = nostrService;
+};
+
+NoscryptSigner::~NoscryptSigner()
+{
+    NCDestroyContext(this->_noscryptContext.get());
+};
+
+#pragma endregion
+
+#pragma region Public Interface
+
+void NoscryptSigner::receiveConnection(string connectionToken)
+{
+    if (connectionToken.empty())
+    {
+        PLOG_ERROR << "No connection token was provided - unable to connect to a remote signer.";
+        return;
+    }
+    
+    int queryStart = this->_parseRemotePublicKey(connectionToken);
+    if (queryStart == -1)
+    {
+        return;
+    }
+
+    string remainingToken = connectionToken.substr(queryStart);
+    int splitIndex = remainingToken.find('&');
+    do
+    {
+        string param = remainingToken.substr(0, splitIndex);
+        this->_handleConnectionTokenParam(param);
+
+        remainingToken = remainingToken.substr(splitIndex + 1);
+        splitIndex = remainingToken.find('&');
+    } while (splitIndex != string::npos);
+
+    // TODO: Handle any messaging with the remote signer.
+};
+
+string NoscryptSigner::initiateConnection(
+    vector<string> relays,
+    string name,
+    string url,
+    string description
+)
+{
+    // Return an empty string if the local keypair is invalid.
+    if (this->_getLocalPrivateKey().empty() || this->_getLocalPublicKey().empty())
+    {
+        PLOG_ERROR << "A valid local keypair is required to connect to a remote signer.";
+        return string();
+    }
+
+    // Return an empty string if no relays are provided.
+    if (relays.empty())
+    {
+        PLOG_ERROR << "At least one relay must be provided to connect to a remote signer.";
+        return string();
+    }
+
+    // Store the provided relay list to be used for sending and receving connection events.
+    this->_relays = relays;
+
+    // Generate the connection token.
+    stringstream ss;
+    ss << "nostrconnect://" << this->_localPublicKey;
+    for (int i = 0; i < relays.size(); i++)
+    {
+        ss << (i == 0 ? "?" : "&");
+        ss << "relay=" << relays[i];
+    }
+    ss << "&metadata={";
+    ss << "\"name\":\"" << name << "\",";
+    ss << "\"url\":\"" << url << "\",";
+    ss << "\"description\":\"" << description << "\"";
+    ss << "}";
+
+    return ss.str();
+
+    // TODO: Handle any messaging with the remote signer.
+};
+
+shared_ptr<promise<bool>> NoscryptSigner::sign(shared_ptr<Event> event)
+{
+    auto signingPromise = make_shared<promise<bool>>();
+
+    auto signerAvailable = this->_pingSigner().get_future();
```

Comment:
Will this synchronously block and is that intended?

Reply:
Yes, that is intended.  We want to make sure the signer is active before asking it to sign an event.

Reply:
It's entirely possible I'm misunderstanding the proper use of futures, here.  I'll research it more and change it if need be.